### PR TITLE
[CBRD-24702] Add sampling query to get Number of Distinct Values.

### PR DESCRIPTION
**Pull request by shparkcubrid**
http://jira.cubrid.org/browse/CBRD-24702

OPTIMIZER
- add 'SAMPLING_SCAN' sql hint

EXECUTOR
- calculate sampling weight (total_page / sampling_page) (scan_open_heap_scan() )
- add sampling heap scan. skip pages by sampling weight (heap_next_sampling () )
- When aggregation is performed, the sample data is multiplied by the sample weight. (qdata_finalize_aggregate_list() )

### CODE REVIEW

File: src/parser/type_checking.c [hunk 1]
```diff
@@ -6797,6 +6797,44 @@ pt_product_sets (PARSER_CONTEXT * parser, TP_DOMAIN * domain, DB_VALUE * set1, D
...
 }
 
+/*
+ * pt_do_where_type () -
+ *   return:
+ *   parser(in):
+ *   node(in):
+ *   arg(in):
+ *   continue_walk(in):
+ */
```
Comment (beyondykk9) [inline src/parser/type_checking.c]: It needs comment to describe the purpose of this routine.

File: src/parser/type_checking.c [hunk 2]
```diff
@@ -20363,6 +20401,20 @@ pt_semantic_type (PARSER_CONTEXT * parser, PT_NODE * tree, SEMANTIC_CHK_INFO * s
       tree = NULL;
     }
 
+  /* When qo_reduce_equality_terms is executed in mq_optimize, a removable predicate like '1=1' is generated.
```
Comment (beyondykk9) [inline src/parser/type_checking.c]: This routine looks not related to this issue. It's better to merge from develop branch.

File: src/parser/type_checking.c [hunk 3]
```diff
@@ -6797,6 +6797,44 @@ pt_product_sets (PARSER_CONTEXT * parser, TP_DOMAIN * domain, DB_VALUE * set1, D
...
+  PT_NODE *spec = NULL;
+
+  if (node == NULL)
+    {
+      return NULL;
+    }
+
+  switch (node->node_type)
+    {
+    case PT_SELECT:
```
Comment (ctshim) [inline src/parser/type_checking.c]: It's trivial, but I think **if(node->node_type == PT_SELET)** is better if you only check PT_SELECT.

**Review by ctshim — APPROVED** [review]

File: src/parser/type_checking.c (hunk 1 above)
Comment (shparkcubrid) [reply src/parser/type_checking.c]: It is not related to this issue.

File: src/parser/type_checking.c (hunk 2 above)
Comment (shparkcubrid) [reply src/parser/type_checking.c]: yes. I did

File: src/parser/type_checking.c (hunk 3 above)
Comment (shparkcubrid) [reply src/parser/type_checking.c]: It is not related to this issue.

**Review by beyondykk9 — APPROVED** [review]

File: src/query/scan_manager.c [hunk 4]
```diff
@@ -2844,6 +2846,19 @@ scan_open_heap_scan (THREAD_ENTRY * thread_p, SCAN_ID * scan_id,
   hsidp->cache_recordinfo = cache_recordinfo;
   hsidp->recordinfo_regu_list = regu_list_recordinfo;
 
+  /* for scampling statistics. */
+  if (scan_type == S_HEAP_SAMPLING_SCAN && !is_partition_table)
+    {
+      int total_pages = 0;
+      if (file_get_num_total_user_pages (thread_p, cls_oid, &total_pages) != NO_ERROR)
```
Comment (youngjinj) [inline src/query/scan_manager.c:2853]: Calls the file_get_num_total_user_pages function for non-partition table cases.
However, the file_get_num_total_user_pages function executes the partition_get_partition_oids function to obtain the number of OIDs of the partition.

Can the function file_get_num_total_user_pages be used for something else?

**Review by youngjinj — APPROVED** [review]

**Review by hyunikn — APPROVED** [review]

**Review by hgryoo — APPROVED** [review]
